### ISSUE TITLE
Pool: Bug with extranonce length panic

### ISSUE THREAD

**Issue by ccdle12** [stratum-mining/stratum#317]
## Overview

I was testing with the Template Provider and found the Pool panics on deserializing a coinbase transaction. It seems to be due to missing an extra 16 bytes in the extra nonce.

Below are some log prints:

- 116 bytes coinbase serialized, taken from here: https://github.com/stratum-mining/stratum/blob/main/protocols/v2/roles-logic-sv2/src/job_creator.rs#L85

    - *DEBUG: serialized coinbase: [2, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 255, 255, 255, 255, 34, 1, 19, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 255, 255, 255, 255, 1, 0, 242, 5, 42, 1, 0, 0, 0, 22, 0, 20, 83, 18, 96, 170, 42, 25, 158, 34, 140, 83, 125, 250, 66, 200, 43, 234, 44, 124, 31, 77, 0, 0, 0, 0]*

- The  prefix, suffix and extranonce are from here (before passing to `merkle_root_from_path()`): https://github.com/stratum-mining/stratum/blob/main/roles/v2/pool/src/lib/mining_pool/mod.rs#L54

    - *DEBUG: prefix: [2, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 255, 255, 255, 255, 34, 1, 19, 0, 0]
    DEBUG: suffix: [255, 255, 1, 0, 242, 5, 42, 1, 0, 0, 0, 22, 0, 20, 83, 18, 96, 170, 42, 25, 158, 34, 140, 83, 125, 250, 66, 200, 43, 234, 44, 124, 31, 77, 0, 0, 0, 0]
    DEBUG: extranonce: [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1]*

- 100 bytes coinbase transaction - printed out here before `Transaction::deserialize(coinbase)`: https://github.com/stratum-mining/stratum/blob/main/protocols/v2/roles-logic-sv2/src/utils.rs#L74

    - *DEBUG: Coinbase bytes that causes failure: [2, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 255, 255, 255, 255, 34, 1, 19, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 255, 255, 1, 0, 242, 5, 42, 1, 0, 0, 0, 22, 0, 20, 83, 18, 96, 170, 42, 25, 158, 34, 140, 83, 125, 250, 66, 200, 43, 234, 44, 124, 31, 77, 0, 0, 0, 0]*

- This leads to the function `Transaction::deserialize(coinbase)` called from inside of `merkle_root_from_path()` to return None since it's not a valid tx and/or missing bytes: https://github.com/stratum-mining/stratum/blob/main/protocols/v2/roles-logic-sv2/src/utils.rs#L74

- The unwrap causes the panic here: https://github.com/stratum-mining/stratum/blob/main/roles/v2/pool/src/lib/mining_pool/mod.rs#L54

## To replicate the bug

1. Git clone `https://github.com/ccdle12/bitcoin/` and checkout `2022.10.18-debug-disable-segwit-outputs` 

2. Build the TP branch:

```
./autogen.sh && ./configure --enable-template-provider
make -j "$(($(nproc)+1))"
```

3. In a shell, start bitcoind in regtest

```
./src/bitcoind -regtest
```

4. In another shell, generate 16 blocks:

```
./src/bitcoin-cli generatetoaddress 16 mrrTnp5a27RRvKXghzs8VAsPfBmWjquTRp
```
5. Start the MVP1 project - git clone `https://github.com/stratum-mining/stratum` and checkout `main`

6. In another shell start the pool

```
cd examples/sv2-proxy && cargo run --bin pool
```
7. In another shell start the translator

```
cd roles/translator && cargo run
```
8. There should be a panic here in the pool

```
thread 'tokio-runtime-worker' panicked at 'called `Option::unwrap()` on a `None` value', roles/v2/pool/src/lib/mining_pool/mod.rs:60:10
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

## Last known working commit

The last known working commit on `main` was: 79fd76699d58513c627eead9ef36ba91a9d262f0

Repeat the steps above but instead of leaving the pool on `main`, run `git checkout 79fd76699d58513c627eead9ef36ba91a9d262f0`

1. Repeat all the steps above to start the pool and translator

2. In another shell start the miner

```
cd examples/sv1-mining-device && cargo run
```
3. In another shell check that empty blocks should be mining

```
[nix-shell:~/Desktop/projects/bitcoin]$ ./src/bitcoin-cli getblockchaininfo
{
  "chain": "regtest",
  "blocks": 97,
  "headers": 97,
  "bestblockhash": "197b4d997417c610601b4865d127ad543c3163bc17ed2c39dcbec89221be66c1",
  "difficulty": 4.656542373906925e-10,
  "time": 1666127144,
  "mediantime": 1666127135,
  "verificationprogress": 1,
  "initialblockdownload": false,
  "chainwork": "00000000000000000000000000000000000000000000000000000000000000c4",
  "size_on_disk": 25019,
  "pruned": false,
  "warnings": "This is a pre-release test build - use at your own risk - do not use for mining or merchant applications"
}
```

**Comment by coleFD** [comment]
I am new to stratum v2, so I am still learning the code base, but I was able to replicate and find a fix for successfully opening the extended mining channel. I do still receive the `Job Id Mismatch` error defined here, but this should be unrelated (https://discordapp.com/channels/950687892169195530/958814900770205739/1044364621374574724) 

For my test, it looks like `extended` in `handle_open_extended_mining_channel()` (https://github.com/stratum-mining/stratum/blob/main/roles/v2/pool/src/lib/mining_pool/message_handler.rs#L185) only gets set to the min length defined in `proxy_config.rs` (16 bytes). By by resizing the extranonce to 32 bytes strictly for the job (`partial_job`), I was able to fix the error without incurring any further downstream errors. Note that even though it gets resized here, it is still resized back to 16 bytes either way for storing prefixes in `self` as well as the `OpenExtendedMiningChannelSuccess` message.

**Comment by ccdle12** [comment]
> I am new to stratum v2, so I am still learning the code base, but I was able to replicate and find a fix for successfully opening the extended mining channel. I do still receive the `Job Id Mismatch` error defined here, but this should be unrelated (https://discordapp.com/channels/950687892169195530/958814900770205739/1044364621374574724)
> 
> For my test, it looks like `extended` in `handle_open_extended_mining_channel()` (https://github.com/stratum-mining/stratum/blob/main/roles/v2/pool/src/lib/mining_pool/message_handler.rs#L185) only gets set to the min length defined in `proxy_config.toml` (16 bytes). By by resizing the extranonce to 32 bytes strictly for the job creation (`partial_job`), I was able to fix the error without incurring any further downstream errors. Note that even though it gets resized here, it is still resized back to 16 bytes either way for storing prefixes in `self` as well as the `OpenExtendedMiningChannelSuccess` message.

Some of these bugs popping up might also be related to this https://github.com/stratum-mining/stratum/issues/340

**Comment by pavlenex** [comment]
This one was addressed in #368
